stable_marriage/gale_shapley.nim

```diff
@@ -0,0 +1,274 @@
+  ## * `contenderPrefs`: recipient ids in descending order of preference
+  ## * `recipientPrefs`: preference score for each contender's id
+  ##
+  assert recipientPrefs.len == recipientPrefs[0].len
+  assert contenderPrefs.len == contenderPrefs[0].len
+  assert recipientPrefs.len == contenderPrefs.len
+  let rosterLen = recipientPrefs.len
+  var
+    # Initializing result sequences with -1, meaning "unmatched"
+    recMatches = newSeqWith(rosterLen, -1)
+    contMatches = newSeqWith(rosterLen, -1)
+    # Queue holding the currently considered "preference score"
+    # (idx of contenderPrefs) for each contender.
+    contQueue = newSeqWith(rosterLen, 0)
+  template match(c, r: Natural) =
+    # Recipient accepts contender, the match is stored
+    contMatches[c] = r
+    recMatches[r] = c
+  while contMatches.contains(-1): # While exist unmatched contenders...
+    for c in 0..<rosterLen: # for each contender...
+      if contMatches[c] == -1: # if they are yet unmatched...
+        # Proposing to the queued recipient
+        let r = contenderPrefs[c][contQueue[c]]
+        inc(contQueue[c]) # Queue next preferred recipient for this contender
+        let rivalMatch = recMatches[r] # Recipient's match index or -1 (vacant)
+        if rivalMatch == -1: # Recipient is also unmatched
+          match(c, r)
+        # Recipient is matched, but contender is more preferable
+        elif recipientPrefs[r][c] < recipientPrefs[r][rivalMatch]:
+          contMatches[rivalMatch] = -1 # Vacate current recipient's match
+          match(c, r)
+        # else: contender's proposition is rejected
+  Matches(contenderMatches: contMatches, recipientMatches: recMatches)
+
+
+func checkMatchingStability*(matches: Matches; contenderPrefs, recipientPrefs:
+                                openArray[Ranking]): Option[(Clash, Clash)] =
+  ## Checks if any matched pair in the current matching is unstable,
+  ## i.e. for **both** of the pair there is no alternative matches preferred
+  ## to the current match.
+  for c, curMatch in matches.contenderMatches.pairs(): # For each contender...
+    # Preference score for the current match
+    let curMatchScore = contenderPrefs[c].find(curMatch)
+    # Try every recipient with higher score, if any
+    for preferredRec in 0..<curMatchScore:
+      # Recipient to check against
+      let checkedRec = contenderPrefs[c][preferredRec]
+      # Current match of the checked recipient
+      let checkedRival = matches.recipientMatches[checkedRec]
+      # If checkedRival's score is worse (>) than contender's score
+      if recipientPrefs[checkedRec][checkedRival] > recipientPrefs[checkedRec][c]:
+        let clashC = (id: c.Natural, match: checkedRec.Natural,
+                      prefers: curMatch.Natural)
+        let clashR = (id: checkedRec.Natural, match: c.Natural,
+                      prefers: checkedRival.Natural)
+        return some((clashC, clashR))
+  none((Clash, Clash))
+
+
+when isMainModule:
```

Comment:
I would add some simple edge cases there:
- I actually think that algorithms have been designed for cases when contenders.len < recipients.len or vice-versa.
It might be interesting to add one example to check an assertion has been raised.

Reply:
Huh, turns out, the implementation is structured in a way where it's not so easy to pass invalid data in. Any suggestions?

Reply:
`doAssertRaises( (name of the defect))`
I can't remember the exact syntax since it does not appear in the docs of std/unittest anymore.

Reply:
Sorry, I was unclear. I wanted that you keep your implementation with the assumption that the sets are of the same length and just add an example with a doAssertRaises, to tell that your implementation catches erroneous inputs.
I think this is actually enough and will accept it like this.